ndmg/stats/qa_tensor.py

```diff
@@ -33,86 +34,167 @@
 import sys
 import matplotlib
 
+from dipy.viz import window, actor
+from fury.actor import orient2rgb
+
 matplotlib.use("Agg")  # very important above pyplot import
 import matplotlib.pyplot as plt
 
-
-def tensor2fa(tensors, tensor_name, dti, derivdir, qcdir):
+def generate_3_d_directions(peak_dirs,peak_values):
     """
-    outdir: location of output directory.
-    fname: name of output fa map file. default is none (name created based on
-    input file)
+    Generates 3-d data required for plotting directions for the entire brain volume
+    
+    Parameters
+    -----------
+    peak_dirs: np array of peak_dirs 
+    peak_values: np array of peak_values
+    slices: dictionary of lists of slices in x,y,z plane
+    
+    Returns
+    -----------
+    centers: np array of center points
+    directions: np array of vector directions
+    directions_colors: np array of RGB colors 
+    heights: np array of vector magnitudes
     """
-    dti_data = nb.load(dti)
-    affine = dti_data.get_affine()
-    dti_data = dti_data.get_data()
-
-    # create FA map
-    FA = fractional_anisotropy(tensors.evals)
-    FA[np.isnan(FA)] = 0
-
-    # generate the RGB FA map
-    FA = np.clip(FA, 0, 1)
-    RGB = color_fa(FA, tensors.evecs)
-
-    fname = os.path.split(tensor_name)[1].split(".")[0] + "_fa_rgb.nii.gz"
-    fa = nb.Nifti1Image(np.array(255 * RGB, "uint8"), affine)
-    nb.save(fa, os.path.join(derivdir, fname))
-
-    fa_pngs(fa, fname, qcdir)
-
-
-def fa_pngs(data, fname, outdir):
+    
+    #initialize 
+    centers = []
+    directions = []
+    heights = []
+    directions_colors = []
+    
+    #iteratively create required new arrow entries per value in 3-d peak object
+    for x in range(peak_dirs.shape[0]):
+        for y in range(peak_dirs.shape[1]): 
+            for z in range(peak_dirs.shape[2]):
+                centers.append([x,y,z])
```

Comment:
I think you can replace this with `[itertools.product](https://docs.python.org/3.7/library/itertools.html#itertools.product)`, which produces the cartesian product of its arguments. Might be a bit cleaner.

so it would be something like
```
xs = range(peak_dirs.shape[0])
ys = range(peak_dirs.shape[1])
zs = range(peak_dirs.shape[2])
for x, y, z in itertools.product(xs, ys, zs):
```

Reply:
There's also a good chance you can do everything within the context of `numpy` without having to loop at all, but I think you're fine, since this code works and it should only be a minor optimization in the grand scheme of things

Reply:
here is some untested code that you may have to fiddle with:

```
coords = np.stack(np.meshgrid(*peak_dirs.shape[:3], indexing='ij'), axis=0)
centers = coords.reshape(-1, 3)
directions = peak_dirs[*coords, 0, :]
directions = directions.reshape(coords.size/3, peak_dirs.shape[-1])
heights = peak_values[*coords, 0].ravel()
```

Reply:
That's _my_ untested code n_n

Reply:
If it makes that much of a difference. most of that is simply to just generate a the cartesian coordinate for the 3-d plotting of the arrows. Otherwise if we remove that 3-d functionality the function reduces to a series of reshapes as i do not need the cartesian coordinate.